### PR TITLE
[3/n] tlsmux: add connection sniffer

### PR DESCRIPTION
**Pull request by jronak**
This PR adds a connection sniffer, which is used in the following PR to sniff the initial bytes from the connection to identify the TLS connection.

### CODE REVIEW

File: transport/internal/tlsmux/conn_sniffer.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		}
+		return n, nil
+	}
+
+	n, err := c.Conn.Read(b)
+	if err != nil {
+		return 0, err
+	}
+
+	if c.sniffing {
```
Comment (gopalchouhan) [inline transport/internal/tlsmux/conn_sniffer.go:53]: Currently we rely on the caller to set sniffing to set true and false explicitly. Can we instead by default set it to true and then toggle it here after storing the buffer?

Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go:53]: I don't have a firm opinion on this, but we can have a constructor method that sets `sniffing = true`. Though my idea was that the creator should control when sniffing should be enabled or disabled. Generically sniffing can be used more than once (maybe capture TLS renegotiation records).

File: transport/internal/tlsmux/conn_sniffer.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+// connSniffer wraps the connection and enables tlsmux to sniff inital bytes from the
+// connection efficiently.
+type connSniffer struct {
+	net.Conn
+
+	sniffing bool         // true when sniffing.
+	buf      bytes.Buffer // holds sniffed data.
+}
+
+func (c *connSniffer) Read(b []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go:44]: Done

File: transport/internal/tlsmux/conn_sniffer.go (hunk 1 above)
Comment (gopalchouhan) [reply transport/internal/tlsmux/conn_sniffer.go:60]: I am not sure if it becomes complicated if we allow multiple sniffing since both the connection might have their own implementation and it could be an inconsistent behaviour in switching the connection unless the bytes are read fully from the connection. Wdyt?

Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go]: At the moment, we don't need in-stream sniffing, so I agree. I have updated the method to sniff by default.

**Review by gopalchouhan — APPROVED** [review]
LGTM

File: transport/internal/tlsmux/conn_sniffer.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	// buf stores bytes read from the underlying connection when in sniffing
+	// mode. When sniffing mode is disabled, buffered bytes is returned.
+	buf bytes.Buffer
+}
+
+// Read returns bytes read from the underlying connection. When sniffing is
+// true, data read from the connection is stored in the buffer. When sniffing
+// mode is disabled, data is first read from the buffer and once the buffer is
+// empty the underlying connection is read.
+func (c *connSniffer) Read(b []byte) (int, error) {
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer.go:44]: It seems like `c.Conn.Read` is destructive, i.e you'll get the first 5 bytes only once.
If that is true, then `*connSniffer.Read` should `defer c.stopSniffing()` at some point, to avoid being misused by the caller.

File: transport/internal/tlsmux/conn_sniffer.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	if c.disableSniffing && c.buf.Len() != 0 {
+		// Read from the buffer when sniffing is disabled and buffer is not empty.
+		n, _ := c.buf.Read(b)
+		if c.buf.Len() == 0 {
+			// Release memory as we don't need buffer anymore.
+			c.buf = bytes.Buffer{}
+		}
+		return n, nil
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer.go:52]: How does this work if the caller wants to read more than 5 bytes? (i.e. `len(b) > c.buff.Len()`)

Don't you need to keep reading from `c.Conn` even after reading `c.buf`?

File: transport/internal/tlsmux/conn_sniffer.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	// mode. When sniffing mode is disabled, buffered bytes is returned.
+	buf bytes.Buffer
+}
+
+// Read returns bytes read from the underlying connection. When sniffing is
+// true, data read from the connection is stored in the buffer. When sniffing
+// mode is disabled, data is first read from the buffer and once the buffer is
+// empty the underlying connection is read.
+func (c *connSniffer) Read(b []byte) (int, error) {
+	if c.disableSniffing && c.buf.Len() != 0 {
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer.go:45]: What happens if you read fewer than 5 bytes from `connSniffer`?
Does `c.buf.Len()` reduce by the number of bytes read?

File: transport/internal/tlsmux/conn_sniffer_test.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func (m *mockConn) Read(b []byte) (int, error) {
+	return m.buf.Read(b)
+}
+
+func TestConnSniffer(t *testing.T) {
+	t.Run("must_read_directly_when_not_sniffing", func(t *testing.T) {
+		data := []byte("test")
+		sniffer := &connSniffer{Conn: &mockConn{buf: bytes.NewBuffer(data)}, disableSniffing: true}
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer_test.go:44]:
```suggestion
		sniffer := &connSniffer{
			Conn: newMockConn(data),
			disableSniffing: true,
		}
```

The person reviewing the code will be more likely to see the important `disableSniffing: true` if it is on its own line.

File: transport/internal/tlsmux/conn_sniffer_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type mockConn struct {
+	net.Conn
+	buf *bytes.Buffer
+}
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer_test.go:35]: The repeated and nested curly braces make the code harder to understand.
Reduce boiler plate in tests by providing a constructor:

```suggestion
}

func newMockConn(data []byte) *mockConn {
	return &mockConn{buf: bytes.NewBuffer(data)}
}
```

File: transport/internal/tlsmux/conn_sniffer_test.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+		data := []byte("test")
+		sniffer := &connSniffer{Conn: &mockConn{buf: bytes.NewBuffer(data)}}
+		require.False(t, sniffer.disableSniffing, "unexpected sniffing value")
+
+		buf := make([]byte, 2)
+		n, err := sniffer.Read(buf)
+		require.NoError(t, err, "unexpected error")
+		assert.Equal(t, 2, n, "unexpected length")
+		assert.Equal(t, data[:2], buf, "unexpected data")
+		assert.Equal(t, data[:2], sniffer.buf.Bytes(), "unexpected buffer capacity")
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer_test.go:64]:
```suggestion
		assert.Equal(t, data[:2], sniffer.buf.Bytes(), "unexpected buffer content")
```
I think you're testing what is currently stored and unread in the buffer, not how much it can hold.

File: transport/internal/tlsmux/conn_sniffer_test.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+	t.Run("must_empty_buffer_after_sniffing", func(t *testing.T) {
+		data := []byte("test")
+		sniffer := &connSniffer{
+			Conn:            &mockConn{buf: bytes.NewBuffer(data[2:])},
+			disableSniffing: true,
+			buf:             *bytes.NewBuffer(data[:2]),
+		}
+
+		sniffer.stopSniffing()
+		require.True(t, sniffer.disableSniffing, "unexpected sniffing value")
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer_test.go:82]: `sniffer.disableSniffing` starts as `true` on line 77. This assertion would still pass if `stopSniffing` did nothing.

File: transport/internal/tlsmux/conn_sniffer_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+		assert.Equal(t, data[2:], buf, "unexpected data")
+		assert.Equal(t, data, sniffer.buf.Bytes(), "unexpected buffer capacity")
+	})
+
+	t.Run("must_empty_buffer_after_sniffing", func(t *testing.T) {
+		data := []byte("test")
+		sniffer := &connSniffer{
+			Conn:            &mockConn{buf: bytes.NewBuffer(data[2:])},
+			disableSniffing: true,
+			buf:             *bytes.NewBuffer(data[:2]),
```
Comment (jkanywhere) [inline transport/internal/tlsmux/conn_sniffer_test.go:78]: t would be a more convincing test if you start with an empty `connSniffer.buf` and call `connSniffer.Read` like you do in previous test cases.

Setting `connSniffer.buf` explicitly in the test feels makes me unsure if it represents a realistic state that `connSniffer` can actually reach.

**Comment by jkanywhere** [comment]
All your `connSniffer` tests use `"test"` as the data. I would appreciate some tests with longer data and the second call to `Read` being for more bytes or fewer bytes than  the first call to `Read`.


optional: Since  the  purpose of `tlsmux.connSniffer` is to check for TLS,  I think we should encapsulate that functionality:
```
func (c *connSniffer) isTLS() (bool, error) {
	return isTLSClientHelloRecord(c)
}
```
That is more of a stylistic choice and does not change the functionality.

File: transport/internal/tlsmux/conn_sniffer.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+		if c.buf.Len() == 0 {
+			// Release memory as we don't need buffer anymore.
+			c.buf = bytes.Buffer{}
+		}
+		return n, nil
+	}
+
+	n, err := c.Conn.Read(b)
+	if err != nil {
+		return 0, err
```
[the remark this answers is not in the document]
Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go:57]: Good call!

File: transport/internal/tlsmux/conn_sniffer.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+	"net"
+)
+
+// connSniffer wraps the connection and enables tlsmux to sniff inital bytes from the
+// connection efficiently.
+type connSniffer struct {
+	net.Conn
+
+	// set to true when sniffing mode is disabled.
+	disableSniffing bool
```
[the remark this answers is not in the document]
Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go:34]: No, there is no concurrent usage of this method (similar to `bytes.Buffer`, `Read` method is lock-free, and the user is supposed to take care of concurrency. 

>Would you ever disable sniffing before calling Read?

I don't think we will disable sniffing before Read, though I prefer keeping a verb in `disableSniffing` to convey that sniffing is disabled currently.

File: transport/internal/tlsmux/conn_sniffer.go (hunk 3 above)
Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go:44]: As this is a private sniffer, IMO, when to stop sniffing must be controlled by the muxer as it holds the sequence of sniffing - parse TLS header and stop sniffing. The number of bytes read (5 bytes) must be unknown to the sniffer.

File: transport/internal/tlsmux/conn_sniffer.go (hunk 5 above)
Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer.go:45]: Yes - https://pkg.go.dev/bytes#Buffer.Len
> Len returns the number of bytes of the unread portion of the buffer; b.Len() == len(b.Bytes()).

File: transport/internal/tlsmux/conn_sniffer_test.go (hunk 9 above)
Comment (jronak) [reply transport/internal/tlsmux/conn_sniffer_test.go:82]: My bad, thanks!
